src/widgets/widget_list/widget_list_item.rs

```diff
@@ -0,0 +1,31 @@
+use crate::{
+    buffer::Buffer,
+    layout::{Rect, Size},
+    widgets::{BarChart, Gauge, Paragraph, SizeHint, Widget},
+};
+
+pub enum WidgetListItem<'a> {
+    Paragraph(Paragraph<'a>),
+    BarChart(BarChart<'a>),
+    Gauge(Gauge<'a>),
+}
+
+impl<'a> SizeHint for WidgetListItem<'a> {
+    fn size_hint(&self, area: &Rect) -> Size {
+        match self {
+            WidgetListItem::Paragraph(e) => e.size_hint(area),
+            WidgetListItem::BarChart(e) => e.size_hint(area),
+            WidgetListItem::Gauge(e) => e.size_hint(area),
+        }
+    }
+}
+
+impl<'a> Widget for WidgetListItem<'a> {
```

Comment:
I was thinking about this last night too. `StatefulWidget` could possibly be the right place to have the behavior that implements the change in how widgets are rendered when they are selected (rather than adding a new Selectable trait like I'd suggested earlier.

Reply:
> behavior that implements the change in how widgets are rendered when they are selected

I see 2 things that can be changed:
1. Style change which is already implemented.
2. different widget / different data: the user needs to pass the widget / data every time at rendering. which can be done right now. the user must check the selected widget and he is able to change the data / widget.

Could you explain a little more, what problem should the `Selectable trait` solve, if the user must check the selected widget and pass a new widget anyway?

Reply:
>Could you explain a little more, what problem should the Selectable trait solve, if the user must check the selected widget and pass a new widget anyway?

To my mind there are two approaches to doing a widget list:
1. Add a widget for each item, every item being a different widget, and explicitly handle putting a different widget in the list based on knowing which list item is selected. This means that the widget list doesn't really even need to have state as that is outside of the widget. The problem with this approach is that it puts the burden of handling list selection stuff on the caller of the List methods. It couples the rendering just a little too much to the state rather than letting the WidgetList handle how to interpret the state.
2. Add the same widget for each item, with just the data part of the widget being different, and let the widget list tell the widget that it is selected / not selected based on the state.

In option 2, the widget can know how to render a selected item differently than a non selected item. I proposed that this would probably mean making a trait where composite widgets can understand whether they are selected when rendering. E.g.:
```rust
impl Widget for MySelectableWidget {
  fn render(...) {
    let borderColor = if (self.is_selected()) {
      Color::Red
    } else {
      Color:default()
    };
    f.render_widget(
      Paragraph::default(self.content)
        .block(Block::default()
          .borders(Borders::all())
          .border_style(Style::default().fg(borderColor)));
  }
}
```
By implementing a Selectable trait on that, the WidgetList can set the selected field to true / false.

Instead, this is probably better just done by making the widgets that get passed in as ListItems implement StatefulWidget. I was thinking outloud when talking about the Selectable trait (and realized that it wasn't really necessary).

Does that make sense?

Reply:
thank you for the explanation. I think maybe that the trade-off of option 1 is still worth it ffor being able to draw what ever you want, based on the state. (but only maybe :smile: )